File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> d;
int n, k;
int max_cnt;
vector<int> answer;
void getSum(int idx, int sum, vector<int>& selectedX, vector<int>& sumX) {
    if(idx == selectedX.size()) {
        sumX.push_back(sum);
        return;
    }
    for(int i = 0; i < 6; i++) {
        getSum(idx + 1, sum + d[selectedX[idx]][i] , selectedX, sumX);
    }
    return;
}

void combi(int start, vector<int> selectedA) {
    if(selectedA.size() == k) {
        vector<int> selectedB;
        for(int i = 0; i < n; i++) {
            if(find(selectedA.begin(), selectedA.end(), i) == selectedA.end()) {
                selectedB.push_back(i);
            }
        }
        vector<int> sumA;
        vector<int> sumB;
        getSum(0, 0, selectedA, sumA);
        getSum(0, 0, selectedB, sumB);
        
        sort(sumB.begin(), sumB.end());
        int cnt = 0;
        for(int a : sumA) {
            cnt += lower_bound(sumB.begin(), sumB.end(), a) - sumB.begin();
        }
        
        if(cnt > max_cnt) {
            max_cnt = cnt;
            answer = selectedA;
        }
        return;
    }
    for(int i = start + 1; i < n; i++) {
        selectedA.push_back(i);
        combi(i, selectedA);
        selectedA.pop_back();
    }
    return;
}

vector<int> solution(vector<vector<int>> dice) {
    d = dice;
    n = d.size();
    k = n / 2;
    combi(-1, {});
    
    for(int & i : answer) {
        i++;
    }
    return answer;
}

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.java
import java.util.*;

class Solution {
    static int[][] d;
    static int n, k, maxCnt;
    static int[] answer;
    public int[] solution(int[][] dice) {
        d = dice;
        n = dice.length;
        k = n / 2;
        answer = new int[k];
        
        combi(-1, new ArrayList<>());
        return answer;
    }
    
    static void combi(int start, List<Integer> selectedA) {
        if(selectedA.size() == k) {
            List<Integer> selectedB = new ArrayList<>();
            for(int i = 0; i < n; i++) {
                if(!selectedA.contains(i)) {
                    selectedB.add(i);
                }
            }
            List<Integer> sumA = new ArrayList<>();
            List<Integer> sumB = new ArrayList<>();
            getSum(0, 0, selectedA, sumA);
            getSum(0, 0, selectedB, sumB);
            
            Collections.sort(sumB);
            int cnt = 0;
            for(int a : sumA) {
                cnt += lowerBound(sumB, a);
            }
            
            if(cnt > maxCnt) {
                maxCnt = cnt;
                for (int i = 0; i < k; i++) {
                    answer[i] = selectedA.get(i) + 1;
                }
            }
            return;
        }
        for(int i = start + 1; i < n; i++) {
            selectedA.add(i);
            combi(i, selectedA);
            selectedA.remove(selectedA.size() - 1);
        }
    }
    
    static void getSum(int idx, int sum, List<Integer> selectedX, List<Integer> sumX) {
        if (idx == selectedX.size()) {
            sumX.add(sum);
            return;
        }
        int diceIdx = selectedX.get(idx);
        for (int i = 0; i < 6; i++) {
            getSum(idx + 1, sum + d[diceIdx][i], selectedX, sumX);
        }
    }
    
    static int lowerBound(List<Integer> list, int target) {
        int left = 0, 
        right = list.size();
        while (left < right) {
            int mid = (left + right) / 2;
            if (list.get(mid) < target) {
                left = mid + 1;
            } 
            else {
                right = mid;
            }
        }
        return left;
    }
}